### DISCUSSION TITLE
when should I use asyncio-mqtt?

### DISCUSSION THREAD

**Discussion by RangeOfGlitching** [empicano/aiomqtt#138] [Q&A] [answered]
Hello everyone, 
I have a general question (sorry about that)
What is the advantage of  asyncio-mqtt over paho-mqtt and what scenarios should I use it in?

I'm using MQTT to control a mutipule robot; some tasks (called by MQTT) take a while, resulting in missed MQTT messages. Could I solve this problem using asyncio-mqtt?

**Comment by empicano** [comment, marked as answer]
Hi there RangeOfGlitching! 😊

Using _asyncio-mqtt_ vs paho-mqtt is ultimately a question of asynchronous vs synchronous code. If you're working on an asynchronous Python program, use _asyncio-mqtt_, if you're working on a synchronous one, use paho-mqtt.

In case you're not familiar with it: AsyncIO is a form of concurrency, like threading or multiprocessing, with the goal of speeding up your program, and is especially suited if your code is waiting a lot (e.g. on HTTP requests, or MQTT messages). If you want to read more about it, I usually find the [RealPython tutorials](https://realpython.com/async-io-python/) of pretty good quality.

The main advantage of _asyncio-mqtt_ over paho-mqtt is speed. The main disadvantage is that you'll be working with async code, which might be a bit more complicated than working with sync code. Both of those things are not really attributable to _async-mqtt_ itself but simply come from the fact that _async-mqtt_ is an asynchronous library and paho-mqtt a synchronous one. If you're not sure, I'd recommend staying with paho-mqtt.

W.r.t. your question about missed messages: This seems like a problem that should be solvable while staying with paho-mqtt. Missing messages is not expected behavior. Maybe you can find some information in the paho-mqtt issues?

~ Felix

**Reply by RangeOfGlitching** [reply]
Thank you for the information about AsyncIO. I always confuse Multithread, Multiprocess, and AsyncIO. I will dive in and try to figure it out. 
I think there was something wrong with my previous question(apology for my English).  me clearify it again.
The message wasn't missing, it was "Delay". Because the on_message will be called until current task finish.
![C1D69AD1-EE67-41ED-A21A-2788CAB33892](https://user-images.githubusercontent.com/58881058/198338155-9057ae57-eea5-490d-bb77-c50ecde5d00f.jpg)

**Reply by JonathanPlasse** [reply]
If you perform a long task within on_message, the following message will be delayed by the time the task took.
You would need to use threads to avoid this problem.

**Reply by empicano** [reply]
Cool chart, that makes it clear! 😎

If you want to process the second message while the first message is still being processed, then yes, you need some sort of concurrency (because you want to do two things at the same time!). You could achieve this with all three types of concurrency: Multiprocessing, Multithreading, and AsyncIO.

I don't know how your robot works exactly, but my first intuition to solve this would be to have paho-mqtt running in the main thread and to start the robot tasks in new threads.

You _could_ do this with AsyncIO and our _asyncio-mqtt_ library, but for this to work well, you would probably need extra AsyncIO libraries for your robot.

**Comment by RangeOfGlitching** [comment]
Thank you for responding so quickly.  The chart seems to explain better than my description. LOL
I appreciated how clear your explanation was.
Multithreading is your first intuition, why is that?

Nvidia Jetson NX is the onboard computer

**Reply by empicano** [reply]
From what you described, it seems to be the easiest option. Multiprocessing is more "heavy-weight" than threading. And as I said, for AsyncIO, you'd probably need extra async libraries. If you want to know more details, there are [great tutorials comparing all three online](https://realpython.com/python-concurrency/) 😊

**Reply by RangeOfGlitching** [reply]
When you say "extra async libraries".  That meening I need something like aiohttp?

**Reply by empicano** [reply]
If you want to run your robot code asynchronously, the code must be asynchronous. If you use any libraries in your code (e.g. to steer your robot), these libraries must be specifically async. For example, for MQTT, there's a sync library, paho-mqtt, and our async library, asyncio-mqtt.

When you want to send asynchronous requests, then yes, you'd use something like aiohttp or httpx instead of requests.

**Comment by frederikaalund** [comment]
Great question! Thanks to everyone for their detailed answers in this discussion already. 👍 Here is my two cents:

Familiarize yourself with the [distinction between concurrency and parallelism](https://stackoverflow.com/questions/1050222/what-is-the-difference-between-concurrency-and-parallelism). Ask yourself if your use case requires one or the other. Usually, concurrency suffices.

Python offers the following libraries out of the box:
 * `asyncio` (concurrency): Always start here IMHO. Need to run something "in the background"? Use `asyncio.create_task`.
 * `threading` (concurrency due to the GIL in CPython): Use this if you need to "unblock" I/O (e.g., if you to `read`/`write` from some file). Usually, you [use `loop.run_in_executor`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor) from asyncio and don't bother with `threading` directly.
 * `multiprocessing` (parallelism): If you need your python code itself to execute in parallel. E.g., for a compute-intensive numerical computation. If you get to this point, you might as well use another (compiled) language IMHO. Like rust (and then use `pyo3` to bind it to python).

See also this answer on StackOverflow: https://stackoverflow.com/a/52498068/554283

So in my opinion, you just go for asyncio to begin with as soon as you need concurrency. If you need parallelism (simultaneous code execution), use bindings to a compiled language. Again, that's my opinion. If you ask somebody else, you may get another recommendation. 😄 

Thanks again to everybody else for their contributions to this discussion. 👍 

~Frederik

**Reply by RangeOfGlitching** [reply]
what a great explain.
I think you are right, for my task it just need concurrency.
